### PR TITLE
[dtensor] lazy init process groups in device mesh

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96700

This PR adds a private flag to allow process grou lazy initialization, this is
replacing the previous `dim_groups` arg, as no one is using that now

This could help avoid creating process groups when not necessary

Differential Revision: [D44044664](https://our.internmc.facebook.com/intern/diff/D44044664)

### CODE REVIEW

**Comment by wanchaol** [comment]
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44044664).

**Comment by wanchaol** [comment]
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44044664).

**Comment by wanchaol** [comment]
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44044664).

**Comment by wanchaol** [comment]
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44044664).

File: torch/distributed/tensor/parallel/_utils.py [hunk 1]
```diff
@@ -149,4 +149,6 @@ def _create_1d_device_mesh(device_mesh: DeviceMesh, tp_mesh_dim: int = 0) -> Dev
     dim_mesh_1d = pg_ranks_by_dim[torch.any(pg_ranks_by_dim == cur_rank, 1), :]
 
     sub_pg = device_mesh.get_dim_groups()[tp_mesh_dim]
-    return DeviceMesh(device_mesh.device_type, dim_mesh_1d.squeeze(), [sub_pg])
+    mesh_1d = DeviceMesh(device_mesh.device_type, dim_mesh_1d.squeeze(), init_process_groups=False)
+    mesh_1d._dim_groups = [sub_pg]
```
Comment (fduwjj) [inline torch/distributed/tensor/parallel/_utils.py:153]: Can we make override an API?

File: torch/distributed/_tensor/device_mesh.py [hunk 2]
```diff
@@ -100,43 +100,88 @@ class DeviceMesh:
...
+            world_size = get_world_size()
+            if self.mesh.numel() > world_size:
+                raise RuntimeError(
+                    f"Mesh should not be bigger than default world size, but found {self.mesh.numel()} ranks!"
+                )
+
+        # TODO: we should do allgather the mesh tensor to ensure every rank have the same mesh value
+
+        # calculate the coordinates of the current global rank on the mesh
+        rank_coords = (self.mesh == get_rank()).nonzero()
```
Comment (fduwjj) [inline torch/distributed/_tensor/device_mesh.py:160]: For my learning, for 1D device mesh each rank is coordinate right?

**Review by fduwjj — APPROVED** [review]
LGTM

File: torch/distributed/_tensor/device_mesh.py (hunk 2 above)
Comment (XilunWu) [reply torch/distributed/_tensor/device_mesh.py:160]: Assume the 1D device mesh (it can be a submesh) [1,3] on 4 ranks.
Because `nonzero()` returns a 2-D Tensor, so rank 1 has `rank_coords` be `[[0]]` and rank 3 has `[[1]]`.

File: torch/distributed/_tensor/device_mesh.py [hunk 3]
```diff
@@ -100,43 +100,88 @@ class DeviceMesh:
...
-        dim_groups: Optional[List[ProcessGroup]] = None,
+        init_process_groups: bool = True,
     ) -> None:
         self.device_type = device_type
         self.mesh = (
             mesh.detach()
             if isinstance(mesh, torch.Tensor)
             else torch.tensor(mesh, dtype=torch.int)
         )
+        # always try to create default (world) pg, even if it is not initialized
```
[the remark this answers is not in the document]
Comment (XilunWu) [reply torch/distributed/_tensor/device_mesh.py:116]: @fduwjj I think it's as the comment said, we still need `self.rank` (e.g. in test code). We won't have it unless the world pg is created in `self._get_or_create_default_group()`.

**Review by XilunWu — APPROVED** [review]
lgtm but one question:
Does this PR  affect my default value PR?

**Comment by wanchaol** [comment]
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44044664).

File: torch/distributed/_tensor/device_mesh.py (hunk 2 above)
Comment (wanchaol) [reply torch/distributed/_tensor/device_mesh.py]: @fduwjj I think we should revisit how 1d device mesh is constructed from 2d mesh, the current way of doing it is somewhat weird. i.e. for 2-d mesh [[0, 1], [2, 3]], we should generate two 1-D mesh, instead of a single 1-D mesh.

File: torch/distributed/tensor/parallel/_utils.py (hunk 1 above)
Comment (wanchaol) [reply torch/distributed/tensor/parallel/_utils.py:153]: Yeah I would prefer to do so if we really need this, but currently I think we should figure out how to slice a 2-D mesh to a 1-D mesh, if override is needed for the slice, we can definitely expose an API to do it

File: torch/distributed/_tensor/device_mesh.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (wanchaol) [reply torch/distributed/_tensor/device_mesh.py:116]: Yeah basically I think we at least need a world_pg in order to establish an rank identity

[the remark this answers is not in the document]
Comment (fduwjj) [reply torch/distributed/_tensor/device_mesh.py:116]: I mean as long as we don't worry about the extra cost of calling `init_process_groups`, this should be fine.

File: torch/distributed/_tensor/device_mesh.py (hunk 2 above)
Comment (fduwjj) [reply torch/distributed/_tensor/device_mesh.py]: Since now we support sub-mesh. So essentially, we split one 2D mesh into two 1D sub-mesh. And we can decide which one to use based on the current rank. Is this something you are saying?

File: torch/distributed/_tensor/device_mesh.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (wanchaol) [reply torch/distributed/_tensor/device_mesh.py:116]: I am not particular sure if this is gonna be very useful, so making it a private arg for now. One thing I think that might be useful is that a non-initialized mesh can be used for sharding planning stage, so lazy init might be make sense.

File: torch/distributed/_tensor/device_mesh.py (hunk 2 above)
Comment (wanchaol) [reply torch/distributed/_tensor/device_mesh.py]: Yes, so basically TP really only run in two sub meshes in a SPMD manner, i.e. in two hosts separately, it does not communicate across ranks(that's the parent mesh sync across the hosts).

Comment (fduwjj) [reply torch/distributed/_tensor/device_mesh.py]: Do you think it makes sense to have a util func inside `device_mesh.py`  to split one 2D or N-D Mesh into lower dim submesh. Or we already have one?

**Comment by wanchaol** [comment]
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44044664).

File: torch/distributed/_tensor/device_mesh.py (hunk 2 above)
Comment (wanchaol) [reply torch/distributed/_tensor/device_mesh.py]: we don't have that yet. I think we can add one to TP first, we can think of whether to make this an API for device_mesh.py once we found there're more use cases :)

We should carefully maintain the device_mesh API going forward as it might be used internally too

**Comment by wanchaol** [comment]
> lgtm but one question: Does this PR affect my default value PR?

@XilunWu as discussed offline, this shouldn't affect the default value PR
